File: golang/internal/ui/datefield.go
package ui

import (
	"fmt"
	"time"

	"gioui.org/font"
	"gioui.org/widget"

	"biorritmo/internal/gfx"
	"biorritmo/internal/theme"
)

// DateField es un selector de fecha compacto (día/mes/año), cada unidad
// con flechas arriba/abajo para incrementar o decrementar -- el
// equivalente funcional del uiDatePicker nativo que usa el proyecto Ruby.
type DateField struct {
	dayUp, dayDown     widget.Clickable
	monthUp, monthDown widget.Clickable
	yearUp, yearDown   widget.Clickable
}

const (
	dfBoxH    = 26.0
	dfDayW    = 30.0
	dfMonthW  = 30.0
	dfYearW   = 46.0
	dfArrowW  = 16.0
	dfUnitGap = 3.0
	dfBorder  = 1.0
	dfTextSz  = 13.0
)

// Width es el ancho total que ocupa el control.
func (d *DateField) Width() float64 {
	unit := dfArrowW
	return (dfDayW + unit) + dfUnitGap + (dfMonthW + unit) + dfUnitGap + (dfYearW + unit)
}

// Height es el alto del control.
func (d *DateField) Height() float64 { return dfBoxH }

// Layout dibuja el selector en (x, y) para date y devuelve la fecha
// resultante y si cambió en este frame.
func (d *DateField) Layout(c *gfx.Canvas, x, y float64, date time.Time, colors theme.Colors) (time.Time, bool) {
	yy, mm, dd := date.Date()
	result := date
	changed := false

	cx := x
	if d.dayUp.Clicked(c.Gtx) {
		result = addDays(result, 1)
		changed = true
	}
	if d.dayDown.Clicked(c.Gtx) {
		result = addDays(result, -1)
		changed = true
	}
	d.drawUnit(c, cx, y, dfDayW, fmt.Sprintf("%02d", dd), colors, &d.dayUp, &d.dayDown)
	cx += dfDayW + dfArrowW + dfUnitGap

	if d.monthUp.Clicked(c.Gtx) {
		result = addMonths(result, 1)
		changed = true
	}
	if d.monthDown.Clicked(c.Gtx) {
		result = addMonths(result, -1)
		changed = true
	}
	d.drawUnit(c, cx, y, dfMonthW, fmt.Sprintf("%02d", int(mm)), colors, &d.monthUp, &d.monthDown)
	cx += dfMonthW + dfArrowW + dfUnitGap

	if d.yearUp.Clicked(c.Gtx) {
		result = addYears(result, 1)
		changed = true
	}
	if d.yearDown.Clicked(c.Gtx) {
		result = addYears(result, -1)
		changed = true
	}
	d.drawUnit(c, cx, y, dfYearW, fmt.Sprintf("%04d", yy), colors, &d.yearUp, &d.yearDown)

	return result, changed
}

func (d *DateField) drawUnit(c *gfx.Canvas, x, y, boxW float64, text string, colors theme.Colors, up, down *widget.Clickable) {
	borderCol := colors.MarkerLine

	c.FillRect(x, y, boxW, dfBoxH, colors.Bg)
	c.FillRect(x, y, boxW, dfBorder, borderCol)
	c.FillRect(x, y+dfBoxH-dfBorder, boxW, dfBorder, borderCol)
	c.FillRect(x, y, dfBorder, dfBoxH, borderCol)
	c.FillRect(x+boxW-dfBorder, y, dfBorder, dfBoxH, borderCol)

	tw, th := c.MeasureText(text, dfTextSz, font.Normal)
	c.Text(text, x+(boxW-tw)/2, y+(dfBoxH-th)/2, dfTextSz, colors.TextH, font.Normal, gfx.AlignLeft)

	ax := x + boxW
	arrowH := dfBoxH / 2
	c.FillRect(ax, y, dfArrowW, dfBoxH, colors.Bg)
	c.FillRect(ax, y, dfArrowW, dfBorder, borderCol)
	c.FillRect(ax, y+dfBoxH-dfBorder, dfArrowW, dfBorder, borderCol)
	c.FillRect(ax, y, dfBorder, dfBoxH, borderCol)
	c.FillRect(ax+dfArrowW-dfBorder, y, dfBorder, dfBoxH, borderCol)
	c.FillRect(ax, y+arrowH-dfBorder/2, dfArrowW, dfBorder, borderCol)

	triCol := colors.Label
	mx := ax + dfArrowW/2
	c.FillPolygon([][2]float64{{mx - 4, y + arrowH - 5}, {mx + 4, y + arrowH - 5}, {mx, y + arrowH - 11}}, triCol)
	c.FillPolygon([][2]float64{{mx - 4, y + arrowH + 5}, {mx + 4, y + arrowH + 5}, {mx, y + arrowH + 11}}, triCol)

	c.HitRegion(up, ax, y, dfArrowW, arrowH)
	c.HitRegion(down, ax, y+arrowH, dfArrowW, arrowH)
}

func daysInMonth(y int, m time.Month) int {
	return time.Date(y, m+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

func addDays(d time.Time, delta int) time.Time {
	return d.AddDate(0, 0, delta)
}

func addMonths(d time.Time, delta int) time.Time {
	y, m, day := d.Date()
	total := int(m) - 1 + delta
	y += total / 12
	m2 := total % 12
	if m2 < 0 {
		m2 += 12
		y--
	}
	newMonth := time.Month(m2 + 1)
	if dim := daysInMonth(y, newMonth); day > dim {
		day = dim
	}
	return time.Date(y, newMonth, day, 0, 0, 0, 0, time.UTC)
}

func addYears(d time.Time, delta int) time.Time {
	y, m, day := d.Date()
	y += delta
	if dim := daysInMonth(y, m); day > dim {
		day = dim
	}
	return time.Date(y, m, day, 0, 0, 0, 0, time.UTC)
}

File: rust/src/ui/datefield.rs
//! Selector de fecha compacto (día/mes/año), cada unidad con flechas
//! arriba/abajo para incrementar o decrementar -- el equivalente funcional
//! del uiDatePicker nativo que usa el proyecto Ruby. Replica datefield.go
//! de la versión Go.

use time::{Date, Duration, Month};

use crate::gfx::{Align, Canvas, Weight};
use crate::theme::Colors;

const BOX_H: f64 = 26.0;
const DAY_W: f64 = 30.0;
const MONTH_W: f64 = 30.0;
const YEAR_W: f64 = 46.0;
const ARROW_W: f64 = 16.0;
const UNIT_GAP: f64 = 3.0;
const BORDER: f64 = 1.0;
const TEXT_SZ: f64 = 13.0;

pub struct DateField;

impl DateField {
    /// Ancho total que ocupa el control.
    pub fn width() -> f64 {
        (DAY_W + ARROW_W) + UNIT_GAP + (MONTH_W + ARROW_W) + UNIT_GAP + (YEAR_W + ARROW_W)
    }

    /// Alto del control.
    pub fn height() -> f64 {
        BOX_H
    }

    /// Dibuja el selector en (x, y) para date y devuelve la fecha resultante
    /// y si cambió en este frame. `id_prefix` debe ser único por instancia
    /// (p.ej. "birth" o "selected") para que las zonas clicables no choquen.
    pub fn show(
        canvas: &Canvas,
        id_prefix: &str,
        x: f64,
        y: f64,
        date: Date,
        colors: &Colors,
    ) -> (Date, bool) {
        let mut result = date;
        let mut changed = false;
        let mut cx = x;

        let day_text = format!("{:02}", date.day());
        if let Some(delta) = Self::draw_unit(
            canvas,
            &format!("{id_prefix}-day"),
            cx,
            y,
            DAY_W,
            &day_text,
            colors,
        ) {
            result = add_days(result, delta);
            changed = true;
        }
        cx += DAY_W + ARROW_W + UNIT_GAP;

        let month_text = format!("{:02}", u8::from(date.month()));
        if let Some(delta) = Self::draw_unit(
            canvas,
            &format!("{id_prefix}-month"),
            cx,
            y,
            MONTH_W,
            &month_text,
            colors,
        ) {
            result = add_months(result, delta as i32);
            changed = true;
        }
        cx += MONTH_W + ARROW_W + UNIT_GAP;

        let year_text = format!("{:04}", date.year());
        if let Some(delta) = Self::draw_unit(
            canvas,
            &format!("{id_prefix}-year"),
            cx,
            y,
            YEAR_W,
            &year_text,
            colors,
        ) {
            result = add_years(result, delta as i32);
            changed = true;
        }

        (result, changed)
    }

    fn draw_unit(
        canvas: &Canvas,
        id_prefix: &str,
        x: f64,
        y: f64,
        box_w: f64,
        text: &str,
        colors: &Colors,
    ) -> Option<i64> {
        let border_col = colors.marker_line;

        canvas.fill_rect(x, y, box_w, BOX_H, colors.bg);
        canvas.fill_rect(x, y, box_w, BORDER, border_col);
        canvas.fill_rect(x, y + BOX_H - BORDER, box_w, BORDER, border_col);
        canvas.fill_rect(x, y, BORDER, BOX_H, border_col);
        canvas.fill_rect(x + box_w - BORDER, y, BORDER, BOX_H, border_col);

        let (tw, th) = canvas.measure_text(text, TEXT_SZ, Weight::Normal);
        canvas.text(
            text,
            x + (box_w - tw) / 2.0,
            y + (BOX_H - th) / 2.0,
            TEXT_SZ,
            colors.text_h,
            Weight::Normal,
            Align::Left,
        );

        let ax = x + box_w;
        let arrow_h = BOX_H / 2.0;
        canvas.fill_rect(ax, y, ARROW_W, BOX_H, colors.bg);
        canvas.fill_rect(ax, y, ARROW_W, BORDER, border_col);
        canvas.fill_rect(ax, y + BOX_H - BORDER, ARROW_W, BORDER, border_col);
        canvas.fill_rect(ax, y, BORDER, BOX_H, border_col);
        canvas.fill_rect(ax + ARROW_W - BORDER, y, BORDER, BOX_H, border_col);
        canvas.fill_rect(ax, y + arrow_h - BORDER / 2.0, ARROW_W, BORDER, border_col);

        let tri_col = colors.label;
        let mx = ax + ARROW_W / 2.0;
        canvas.fill_polygon(
            &[
                (mx - 4.0, y + arrow_h - 5.0),
                (mx + 4.0, y + arrow_h - 5.0),
                (mx, y + arrow_h - 11.0),
            ],
            tri_col,
        );
        canvas.fill_polygon(
            &[
                (mx - 4.0, y + arrow_h + 5.0),
                (mx + 4.0, y + arrow_h + 5.0),
                (mx, y + arrow_h + 11.0),
            ],
            tri_col,
        );

        let up_clicked = canvas
            .interact_click(ax, y, ARROW_W, arrow_h, format!("{id_prefix}-up"))
            .clicked();
        let down_clicked = canvas
            .interact_click(
                ax,
                y + arrow_h,
                ARROW_W,
                arrow_h,
                format!("{id_prefix}-down"),
            )
            .clicked();

        if up_clicked {
            Some(1)
        } else if down_clicked {
            Some(-1)
        } else {
            None
        }
    }
}

fn days_in_month(year: i32, month: Month) -> u8 {
    let (next_year, next_month) = if month == Month::December {
        (year + 1, Month::January)
    } else {
        (year, month.next())
    };
    let first_of_next = Date::from_calendar_date(next_year, next_month, 1).expect("fecha válida");
    (first_of_next - Duration::days(1)).day()
}

fn add_days(d: Date, delta: i64) -> Date {
    d + Duration::days(delta)
}

fn add_months(d: Date, delta: i32) -> Date {
    let day = d.day();
    let month_index = i32::from(u8::from(d.month())) - 1;
    let total = month_index + delta;
    let year = d.year() + total.div_euclid(12);
    let month = Month::try_from((total.rem_euclid(12) + 1) as u8).expect("mes válido");
    let dim = days_in_month(year, month);
    Date::from_calendar_date(year, month, day.min(dim)).expect("fecha válida")
}

fn add_years(d: Date, delta: i32) -> Date {
    let year = d.year() + delta;
    let month = d.month();
    let dim = days_in_month(year, month);
    Date::from_calendar_date(year, month, d.day().min(dim)).expect("fecha válida")
}
